lms/views/decorators/h_api.py

```diff
@@ -193,25 +162,85 @@ def wrapper(request, jwt):
 
         authority = request.registry.settings["h_authority"]
         userid = f"acct:{username}@{authority}"
-        client_id = request.registry.settings["h_client_id"]
-        client_secret = request.registry.settings["h_client_secret"]
+        url = f"/groups/{group.pubid}/members/{userid}"
 
-        try:
-            response = requests.post(
-                request.registry.settings["h_api_url"]
-                + f"/groups/{group.pubid}/members/{userid}",
-                auth=(client_id, client_secret),
-                timeout=1,
-            )
-            response.raise_for_status()
-        except RequestException:
-            raise HTTPGatewayTimeout(explanation="Connecting to Hypothesis failed")
+        post(request.registry.settings, url)
 
         return wrapped(request, jwt)
 
     return wrapper
 
 
+def post(settings, path, data=None, username=None, statuses=None):
+    """
+    Do an H API post and return the response.
+
+    If the request fails for any reason (for example a network connection error
+    or a timeout) :exc:`~lms.views.HAPIError` is raised.
+    :exc:`~lms.views.HAPIError` is also raised if a 4xx or 5xx response is
+    received. Use the optional keyword argument ``statuses`` to supply a list
+    of one or more 4xx or 5xx statuses for which :exc:`~lms.views.HAPIError`
+    should not be raised -- the 4xx or 5xx response will be returned instead.
+
+    :arg settings: the Pyramid request.registry.settings object
+    :type settings: dict
+    :arg path: the h API path to post to, relative to
+      ``settings["h_api_url"]``, for example: ``"/users"`` or
+      ``"/groups/<PUBID>/members/<USERID>"``
+    :type url: str
+    :arg data: the data to post as JSON in the request body
+    :type data: dict
+    :arg username: the username of the user to post as (using an
+      X-Forwarded-User header)
+    :type username: str
+    :arg statuses: the list of 4xx and 5xx statuses that should not trigger an
+      exception, for example: ``[409, 410]``
+    :type statuses: list of ints
+
+    :raise HAPIError: if the request fails for any reason, including if a 4xx
+      or 5xx response is received
+
+    :return: the response from the h API
+    :rtype: requests.Response
+    """
+    statuses = statuses or []
+
+    # Our OAuth 2.0 client_id and client_secret for authenticating to the h API.
+    client_id = settings["h_client_id"]
+    client_secret = settings["h_client_secret"]
+
+    # The authority that we'll create h users and groups in.
+    authority = settings["h_authority"]
+
+    # The full h API URL to post to.
+    url = settings["h_api_url"] + path
+
+    post_args = dict(url=url, auth=(client_id, client_secret), timeout=1)
```

Comment:
This timeout seems quite short to me. Is that going to be long enough?

Reply:
It seems to have been fine so far, although there seems to be no reason not to increase it

Reply:
Increased it to 10s